File: src/base91x/base91x.py
"""Base91x encoder-decoder
MIT License

Copyright (c) 2022 Roman Babenko

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

# BASE91X_ALPHABET in order
BASE91X_ALPHABET = "!~}|{zyxwvutsrqponmlkjihgfedcba`_^]#[ZYXWVUTSRQPONMLKJIHGFEDCBA@?>=<;:9876543210/.-,+*)($&%"

# Base of the numeric system is 91 dec equals ASCII symbol [
BASE91X_LEN = 91

# Bits in one byte. Should be 8
CHAR_BIT = 8

# Pair of base91x symbols might code 13 bits
BASE91X_WORD_BIT = 13

# 8192 possibly values for 13 bits
BASE91X_WORD_SIZE = 0x2000

# Mask for 13 bits
BASE91X_WORD_MASK = 0x1FFF

# BASE91X reverse table for quick decoding
BASE91X_ZYX = [
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  #
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  #
    -1, 0, -1, 35, 88, 90, 89, -1, 87, 86, 85, 84, 83, 82, 81, 80,  #
    79, 78, 77, 76, 75, 74, 73, 72, 71, 70, 69, 68, 67, 66, 65, 64,  #
    63, 62, 61, 60, 59, 58, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48,  #
    47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, -1, 34, 33, 32,  #
    31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16,  #
    15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, -1]

ZYX_LEN = len(BASE91X_ZYX)

ZYX_MASK = ZYX_LEN - 1


# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

def encode(data: bytes) -> str:
    """Encodes bytes to base91x text

    Args:
        data: input bytes
    Return:
        string with encoded data
    """
    text = ""

    collector: int = 0
    bit_collected: int = 0
    for i in data:
        collector |= i << bit_collected
        bit_collected += CHAR_BIT
        while BASE91X_WORD_BIT <= bit_collected:
            cod = BASE91X_WORD_MASK & collector
            quotient, remainder = divmod(cod, BASE91X_LEN)
            text += BASE91X_ALPHABET[remainder]
            text += BASE91X_ALPHABET[quotient]
            collector >>= BASE91X_WORD_BIT
            bit_collected -= BASE91X_WORD_BIT

    if 0 != bit_collected:
        cod = BASE91X_WORD_MASK & collector
        quotient, remainder = divmod(cod, BASE91X_LEN)
        text += BASE91X_ALPHABET[remainder]
        if 7 <= bit_collected:
            text += BASE91X_ALPHABET[quotient]
    return text


# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

def decode(text: str) -> bytearray:
    """Decodes text with the base91x algorithm, skipping wrong symbols

    Args:
        text: input string with encoded data
    Return:
        decoded bytes
    """

    data = bytearray()
    collector: int = 0
    bit_collected: int = 0
    lower: int = -1
    for symbol in text:
        i = ord(symbol)
        if ZYX_MASK < i:
            continue
        digit = BASE91X_ZYX[ZYX_MASK & i]
        if -1 == digit:
            continue
        if -1 == lower:
            lower = digit
            continue

        collector |= (BASE91X_LEN * digit + lower) << bit_collected
        bit_collected += BASE91X_WORD_BIT
        lower = -1

        while CHAR_BIT <= bit_collected:
            data.append(0xFF & collector)
            collector >>= CHAR_BIT
            bit_collected -= CHAR_BIT

    if -1 != lower:
        collector |= lower << bit_collected
        bit_collected += 7  # CHAR_BIT - 1

    if CHAR_BIT <= bit_collected:
        data.append(0xFF & collector)

    return data

File: src/main/java/base91x/Base91x.java
/*
MIT License

Copyright (c) 2026 Roman Babenko

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
package base91x;

public final class Base91x {

    private Base91x() {}

    // --- constants ---

    private static final int BASE = 91;
    private static final int CHAR_BITS = 8;
    private static final int WORD_BITS = 13;
    private static final int WORD_MASK = 0x1FFF;

    // --- alphabet ---

    private static final byte[] ALPHABET = new byte[] {
        '!', '~', '}', '|', '{', 'z', 'y', 'x', 'w', 'v',
        'u', 't', 's', 'r', 'q', 'p', 'o', 'n', 'm', 'l',
        'k', 'j', 'i', 'h', 'g', 'f', 'e', 'd', 'c', 'b',
        'a', '`', '_', '^', ']', '#', '[', 'Z', 'Y', 'X',
        'W', 'V', 'U', 'T', 'S', 'R', 'Q', 'P', 'O', 'N',
        'M', 'L', 'K', 'J', 'I', 'H', 'G', 'F', 'E', 'D',
        'C', 'B', 'A', '@', '?', '>', '=', '<', ';', ':',
        '9', '8', '7', '6', '5', '4', '3', '2', '1', '0',
        '/', '.', '-', ',', '+', '*', ')', '(', '$', '&',
        '%'
    };

    private static final int[] REVERSE = new int[256];

    static {
        for (int i = 0; i < REVERSE.length; i++) {
            REVERSE[i] = -1;
        }
        for (int i = 0; i < ALPHABET.length; i++) {
            REVERSE[ALPHABET[i] & 0xFF] = i;
        }
    }

    // --- size helpers ---

    public static int computeEncodedSize(int size) {
        size <<= 4;
        return (size % WORD_BITS != 0)
                ? (size / WORD_BITS + 1)
                : (size / WORD_BITS);
    }

    public static int assumeDecodedSize(int size) {
        size *= WORD_BITS;
        size >>= 4;
        return size;
    }

    // --- encode ---

    public static byte[] encode(byte[] input) {
        if (input == null) return null;

        byte[] out = new byte[computeEncodedSize(input.length)];

        int outPos = 0;
        int collector = 0;
        int bits = 0;

        for (byte b : input) {
            collector |= (b & 0xFF) << bits;
            bits += CHAR_BITS;

            while (bits >= WORD_BITS) {
                int value = collector & WORD_MASK;

                int rem = value % BASE;
                int quot = value / BASE;

                out[outPos++] = ALPHABET[rem];
                if (outPos >= out.length) break;

                out[outPos++] = ALPHABET[quot];

                collector >>>= WORD_BITS;
                bits -= WORD_BITS;
            }
        }

        if (bits > 0 && outPos < out.length) {
            int value = collector & WORD_MASK;

            int rem = value % BASE;
            int quot = value / BASE;

            out[outPos++] = ALPHABET[rem];

            if (bits >= 7 && outPos < out.length) {
                out[outPos++] = ALPHABET[quot];
            }
        }

        if (outPos == out.length) {
            return out;
        }

        // trim (Java-style)
        byte[] result = new byte[outPos];
        System.arraycopy(out, 0, result, 0, outPos);
        return result;
    }

    // --- decode ---

    public static byte[] decode(byte[] input) {
        if (input == null) return null;

        byte[] out = new byte[assumeDecodedSize(input.length)];

        int outPos = 0;
        int collector = 0;
        int bits = 0;
        int lower = -1;

        for (byte b : input) {
            int digit = REVERSE[b & 0xFF];
            if (digit == -1) continue;

            if (lower == -1) {
                lower = digit;
                continue;
            }

            collector |= (BASE * digit + lower) << bits;
            bits += WORD_BITS;
            lower = -1;

            while (bits >= CHAR_BITS && outPos < out.length) {
                out[outPos++] = (byte) (collector & 0xFF);
                collector >>>= CHAR_BITS;
                bits -= CHAR_BITS;
            }
        }

        if (lower != -1) {
            collector |= lower << bits;
            bits += (CHAR_BITS - 1);
        }

        if (bits >= CHAR_BITS && outPos < out.length) {
            out[outPos++] = (byte) (collector & 0xFF);
        }

        byte[] result = new byte[outPos];
        System.arraycopy(out, 0, result, 0, outPos);
        return result;
    }
}